### PR TITLE
add support for SST-BF features

### PR DESCRIPTION
**Pull request by rkamudhan**
* add discovery of SST-BF high-priority core through NFD

For more information, please refer to https://www.intel.co.uk/content/www/uk/en/architecture-and-technology/speed-select-technology-article.html

Change-Id: I4b2031ef40d206842d9f6cef188a14a0e47f1dea

### CODE REVIEW

File: cmd/nfd-worker/main.go [hunk 1]
```diff
@@ -320,6 +321,7 @@ func configureParameters(sourcesWhiteList []string, labelWhiteListStr string) (e
 		network.Source{},
 		panic_fake.Source{},
 		pci.Source{},
+		cpu_power.Source{},
```
Comment (marquiz) [inline cmd/nfd-worker/main.go]: ...and, remove...

File: source/cpu/cpu_power/cpupower_amd64.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+const (
+	// CPUID EAX input values
+	PROCESSOR_FREQ_INFO_ENUMERATION_LEAF = 0x16
+)
+
+func DiscoverBaseFreq() uint32 {
+	freqInfo := cpuid_low(PROCESSOR_FREQ_INFO_ENUMERATION_LEAF, 0)
+	return freqInfo.EAX
+}
```
Comment (marquiz) [inline source/cpu/cpu_power/cpupower_amd64.go]: Please do not copy paste cpuid functionality that already exists in `source/rdt/`. But, make it common, instead, and use it from rdt and from cpu-power. That is, move the cpuid code from `rdt` to `pkg/cpuid`, for example.

File: source/cpu/cpu_power/cpupower.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	re := regexp.MustCompile("cpu([0-9])")
+	var i int
+	for _, f := range files {
+		if re.MatchString(f.Name()) {
+			i++
+		}
+	}
+
+	return i, nil
+}
```
Comment (marquiz) [inline source/cpu/cpu_power/cpupower.go]: I don't think you need this function. Just read `/sys/bus/cpu/devices/*`

File: source/cpu/cpu_power/cpupower.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	var pbfenabled bool
+
+	cpucount, err := getcpuCount()
+	if err != nil {
+		return pbfenabled, err
+	}
+
+	P1 := getcpuBaseFrequency()
+
+	for cpuid := 0; cpuid < cpucount; cpuid++ {
```
Comment (marquiz) [inline source/cpu/cpu_power/cpupower.go]: Just do `readDir()` for in `/sys/bus/cpu/devices/` as I pointed above

File: source/cpu/cpu_power/cpupower.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	return i, nil
+}
+
+func getBaseFrequency(cpu int) (int, error) {
+	var baseFreq int
+
+	baseFile := fmt.Sprintf("%s%d%s", sysdevcpu, cpu, cpubasefreq)
+	if _, err := os.Lstat(baseFile); err != nil {
+		return baseFreq, &NoBasefreqSupport{"no base frequency support"}
+	}
```
Comment (marquiz) [inline source/cpu/cpu_power/cpupower.go]: I think you do not need this. Just try to read the file and use `os.IsNotExist()` for catching the case where file does not exist

File: source/cpu/cpu_power/cpupower.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+			return nil, fmt.Errorf("can't detect whether pbf is enabled: %s", err.Error())
+		}
+	}
+
+	if pbfenabled != false {
+		features["sst_bf.enabled"] = true
+	}
+
+	return features, nil
+}
```
Comment (marquiz) [inline source/cpu/cpu_power/cpupower.go]: The `Discover()´ functionality should be moved to `source/cpu/cpu.go`

File: README.md [hunk 7]
```diff
@@ -353,6 +354,12 @@ possible security implications.
 | :----------: | ------------------------------------------------------------- |
 | turbo        | Turbo frequencies are enabled in Intel pstate driver
 
+### System Features
+
+| Feature     | Attribute        | Description                                 |
+| ----------- | ---------------- | --------------------------------------------|
+| cpu_power   | sst_bf.enabled   | Speed Select Technology - Base frequency is enabled in cpu           |
```
Comment (marquiz) [inline README.md]: Move under `cpu` source. I think feature could be `power` and attribute as is `sst_bf.enabled`.

File: cmd/nfd-worker/main.go [hunk 8]
```diff
@@ -37,6 +37,7 @@ import (
 	"sigs.k8s.io/node-feature-discovery/pkg/version"
 	"sigs.k8s.io/node-feature-discovery/source"
 	"sigs.k8s.io/node-feature-discovery/source/cpu"
+	"sigs.k8s.io/node-feature-discovery/source/cpu/cpu_power"
```
[the remark this answers is not in the document]
Comment (rkamudhan) [reply cmd/nfd-worker/main.go]: +1, will change it.

File: cmd/nfd-worker/main.go [hunk 9]
```diff
@@ -218,7 +219,7 @@ func argsParse(argv []string) (Args, error) {
                               in testing
                               [Default: ]
   --sources=<sources>         Comma separated list of feature sources.
-                              [Default: cpu,cpuid,iommu,kernel,local,memory,network,pci,pstate,rdt,storage,system]
+                              [Default: cpu,cpuid,iommu,kernel,local,memory,network,cpu_power,pci,pstate,rdt,storage,system]
```
[the remark this answers is not in the document]
Comment (rkamudhan) [reply cmd/nfd-worker/main.go]: +1, will change it.

File: cmd/nfd-worker/main.go (hunk 1 above)
Comment (rkamudhan) [reply cmd/nfd-worker/main.go]: +1, will change it.

File: source/cpu/cpu_power/cpupower_amd64.go (hunk 2 above)
Comment (rkamudhan) [reply source/cpu/cpu_power/cpupower_amd64.go]: +1, will change it.

File: source/cpu/cpu_power/cpupower.go (hunk 3 above)
Comment (rkamudhan) [reply source/cpu/cpu_power/cpupower.go]: +1, will change it.

File: source/cpu/cpu_power/cpupower.go (hunk 4 above)
Comment (rkamudhan) [reply source/cpu/cpu_power/cpupower.go]: +1, will change it.

File: source/cpu/cpu_power/cpupower.go (hunk 5 above)
Comment (rkamudhan) [reply source/cpu/cpu_power/cpupower.go]: +1, will change it.

File: source/cpu/cpu_power/cpupower.go (hunk 6 above)
Comment (rkamudhan) [reply source/cpu/cpu_power/cpupower.go]: +1, will change it.

File: README.md (hunk 7 above)
Comment (rkamudhan) [reply README.md]: +1, will change it.

**Comment by marquiz** [comment]
Obsoleted by #235

**Pull request closed without merging**